### PR TITLE
refactor: reduce code expand in codegen

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Tests

- [x] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/databend/16961)

### CODE REVIEW

File: src/query/expression/src/new_register.rs [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+pub fn vectorize_2_arg<I1: ArgType, I2: ArgType, O: ArgType>(
+    func: impl Fn(I1::ScalarRef<'_>, I2::ScalarRef<'_>, &mut EvalContext) -> O::Scalar
+    + Copy
+    + Send
+    + Sync,
+) -> impl Fn(Value<I1>, Value<I2>, &mut EvalContext) -> Value<O> + Copy + Send + Sync {
+    move |arg1, arg2, ctx| {
+        let generics = ctx.generics.to_vec();
+        let iter = (0..ctx.num_rows).map(|index| {
+            let arg1 = unsafe { arg1.index_unchecked(index) };
```
Comment (sundy-li) [inline src/query/expression/src/new_register.rs:55]: there will be match case inside the loop, did you check the performance benches?

Comment (sundy-li) [reply src/query/expression/src/new_register.rs:55]:
```
Gnuplot not found, using plotters backend
bench_kernels/register_old/10240
                        time:   [6.9478 µs 6.9775 µs 7.0112 µs]
Found 10 outliers among 100 measurements (10.00%)
  10 (10.00%) high severe
bench_kernels/register_new/10240
                        time:   [3.1606 µs 3.1758 µs 3.1931 µs]
Found 10 outliers among 100 measurements (10.00%)
  10 (10.00%) high severe
bench_kernels/register_old/102400
                        time:   [67.850 µs 68.373 µs 68.960 µs]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe
bench_kernels/register_new/102400
                        time:   [32.901 µs 33.321 µs 33.816 µs]
Found 15 outliers among 100 measurements (15.00%)
```

https://logpaste.com/QjggJAAN

Comment (sundy-li) [reply src/query/expression/src/new_register.rs:55]: I think it's ok to keep old_register for  `1_arg` and `2_arg`.

For multi args we can use new to reduce code expand

Comment (andylokandy) [reply src/query/expression/src/new_register.rs:55]: Seems like the new one is even faster?

Comment (sundy-li) [reply src/query/expression/src/new_register.rs:55]: It's revered typo

**Pull request closed without merging**